cpp/src/io/parquet/experimental/hybrid_scan.cpp

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hybrid_scan_impl.hpp"
+
+#include <cudf/io/experimental/hybrid_scan.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf::io::parquet::experimental {
+
+hybrid_scan_reader::hybrid_scan_reader(cudf::host_span<uint8_t const> footer_bytes,
+                                       cudf::io::parquet_reader_options const& options)
+  : _impl{std::make_unique<detail::impl>(footer_bytes, options)}
+{
+}
+
+hybrid_scan_reader::~hybrid_scan_reader() = default;
+
+[[nodiscard]] text::byte_range_info hybrid_scan_reader::get_page_index_bytes() const
+{
+  return _impl->get_page_index_bytes();
+}
+
+[[nodiscard]] FileMetaData const& hybrid_scan_reader::get_parquet_metadata() const
+{
+  return _impl->get_parquet_metadata();
+}
+
+void hybrid_scan_reader::setup_page_index(cudf::host_span<uint8_t const> page_index_bytes) const
+{
+  return _impl->setup_page_index(page_index_bytes);
+}
+
+std::vector<cudf::size_type> hybrid_scan_reader::get_all_row_groups(
+  cudf::io::parquet_reader_options const& options) const
+{
+  CUDF_EXPECTS(options.get_row_groups().size() == 0 or options.get_row_groups().size() == 1,
+               "Encountered invalid size of row group indices in parquet reader options");
+
+  // If row groups are specified in parquet reader options, return them as is
+  if (options.get_row_groups().size()) { return options.get_row_groups().front(); }
+
+  return _impl->get_all_row_groups(options);
+}
+
+std::vector<cudf::size_type> hybrid_scan_reader::filter_row_groups_with_stats(
+  cudf::host_span<size_type const> row_group_indices,
+  cudf::io::parquet_reader_options const& options,
+  rmm::cuda_stream_view stream) const
+{
+  // Temporary vector with row group indices from the first source
+  auto const input_row_group_indices =
```

Comment:
These APIs simply create a vector of `row_group_indices` with one element for now and pass them on to `_impl`. This is to enable easy expansion of these APIs to multiple input sources in the near future accepting a span of vectors (instead of span of ints) and directly pass it on to `_impl`

Reply:
@vuule currently only supporting single parquet file read in the hybrid scan reader. Will extend to multiple files in the future and remove this temp stuff